File: ios/stacker/Sources/stacker/StackerToast.swift
import UIKit

/// Shows a per-call toast in a **native iOS** host app, styled to match the
/// Flutter dashboard.
///
/// In a Flutter app the toast stack is drawn by `StackerOverlay`. A native
/// host has no Flutter widget tree on screen — the engine only runs while the
/// dashboard is presented — so that overlay can never appear. Without this, a
/// native iOS app captures everything correctly but shows nothing, which reads
/// as "the library is not working".
///
/// The design deliberately mirrors `_ToastCard` and `StackerAndroid`'s
/// `StackerToast`: the same dark surface, accent border, tinted status badge,
/// monospaced type scale and chevron affordance. A developer who sees the
/// Flutter and native versions on one project should not be able to tell them
/// apart.
///
/// Enabled via `StackerAutoAttach.enable`; a no-op until then.
@objc public final class StackerToast: NSObject {

    /// Toast policy, mirroring `ToastPolicy` on the Dart side.
    @objc public enum Policy: Int {
        /// A toast for every completed call.
        case all

        /// Only non-2xx responses and transport failures.
        case errorsOnly

        /// No toasts.
        case none
    }

    // Design tokens copied from StackerTheme. The card paints its own dark
    // surface, so the *dark* variant of each status colour is correct
    // regardless of the host app's appearance.
    private static let surface = UIColor(red: 0.106, green: 0.125, blue: 0.161, alpha: 0.94)
    private static let success = UIColor(red: 0.290, green: 0.871, blue: 0.502, alpha: 1)
    private static let redirect = UIColor(red: 0.655, green: 0.545, blue: 0.980, alpha: 1)
    private static let warn = UIColor(red: 0.984, green: 0.749, blue: 0.141, alpha: 1)
    private static let errorColor = UIColor(red: 0.973, green: 0.443, blue: 0.443, alpha: 1)
    private static let neutral = UIColor(red: 0.580, green: 0.639, blue: 0.722, alpha: 1)

    /// How long a toast stays on screen, matching the Flutter default.
    private static let visibleDuration: TimeInterval = 3.0

    /// Bursts are coalesced so parallel requests cannot cover the screen.
    private static let maxVisible = 3

    private static var policy: Policy = .none
    private static var liveToasts: [UIView] = []

    /// Whether toasts are currently enabled.
    @objc public static var isEnabled: Bool { policy != .none }

    /// Configures the emitter. Called by `StackerAutoAttach.enable`.
    @objc public static func configure(policy: Policy) {
        self.policy = policy
    }

    /// Turns toasts off and removes any on screen.
    @objc public static func disable() {
        policy = .none
        DispatchQueue.main.async { dismissAll() }
    }

    /// Shows a toast for a completed call, if policy allows it.
    ///
    /// - Parameter statusCode: the HTTP status, or `nil` for a transport failure.
    @objc public static func showApiCall(
        method: String,
        path: String,
        statusCode: NSNumber?,
        durationMs: NSNumber?
    ) {
        let code = statusCode?.intValue
        let isError = code == nil || code! >= 400

        switch policy {
        case .none: return
        case .errorsOnly: if !isError { return }
        case .all: break
        }

        let badge = code.map(String.init) ?? "ERR"
        let accent = accentFor(code)
        var subtitle = code == nil ? "transport failure" : "completed"
        if let ms = durationMs?.intValue {
            subtitle += "  •  \(ms) ms"
        }

        onMain {
            show(badge: badge, title: "\(method) \(path)", subtitle: subtitle,
                 accent: accent, tab: "api")
        }
    }

    /// Shows a toast for a captured crash. Always shown when enabled.
    @objc public static func showCrash(errorType: String, fatal: Bool) {
        guard policy != .none else { return }
        onMain {
            show(
                badge: fatal ? "FATAL" : "WARN",
                title: errorType,
                subtitle: "tap to open the dashboard",
                accent: fatal ? errorColor : warn,
                tab: "crashes"
            )
        }
    }

    /// Mirrors `StackerTheme.statusColor` for the dark surface.
    private static func accentFor(_ statusCode: Int?) -> UIColor {
        guard let code = statusCode else { return errorColor }
        switch code {
        case 200..<300: return success
        case 300..<400: return redirect
        case 400..<500: return warn
        case 500...: return errorColor
        default: return neutral
        }
    }

    /// Builds and attaches the card. Main thread only.
    private static func show(
        badge: String,
        title: String,
        subtitle: String,
        accent: UIColor,
        tab: String
    ) {
        guard let window = keyWindow() else { return }

        let card = ToastCard(accent: accent, tab: tab)
        card.translatesAutoresizingMaskIntoConstraints = false
        card.backgroundColor = surface
        card.layer.cornerRadius = 10
        card.layer.borderWidth = 1
        card.layer.borderColor = accent.withAlphaComponent(0.55).cgColor
        card.layer.shadowColor = UIColor.black.cgColor
        card.layer.shadowOpacity = 0.28
        card.layer.shadowRadius = 10
        card.layer.shadowOffset = CGSize(width: 0, height: 3)

        // Status badge.
        let badgeLabel = PaddedLabel()
        badgeLabel.text = badge
        badgeLabel.textColor = accent
        badgeLabel.font = .monospacedSystemFont(ofSize: 11, weight: .bold)
        badgeLabel.backgroundColor = accent.withAlphaComponent(0.16)
        badgeLabel.layer.cornerRadius = 4
        badgeLabel.clipsToBounds = true
        badgeLabel.textAlignment = .center
        badgeLabel.translatesAutoresizingMaskIntoConstraints = false

        let titleLabel = UILabel()
        titleLabel.text = title
        titleLabel.textColor = .white
        titleLabel.font = .monospacedSystemFont(ofSize: 12, weight: .semibold)
        titleLabel.lineBreakMode = .byTruncatingTail
        titleLabel.textAlignment = .left

        let subtitleLabel = UILabel()
        subtitleLabel.text = subtitle
        subtitleLabel.textColor = UIColor.white.withAlphaComponent(0.7)
        subtitleLabel.font = .monospacedSystemFont(ofSize: 10, weight: .regular)
        subtitleLabel.lineBreakMode = .byTruncatingTail
        subtitleLabel.textAlignment = .left

        let textColumn = UIStackView(arrangedSubviews: [titleLabel, subtitleLabel])
        textColumn.axis = .vertical
        textColumn.alignment = .leading
        textColumn.spacing = 1

        let chevron = UILabel()
        chevron.text = "›"
        chevron.textColor = UIColor.white.withAlphaComponent(0.54)
        chevron.font = .systemFont(ofSize: 16, weight: .regular)

        let row = UIStackView(arrangedSubviews: [badgeLabel, textColumn, chevron])
        row.axis = .horizontal
        row.alignment = .center
        row.spacing = 9
        row.translatesAutoresizingMaskIntoConstraints = false

        // The badge and chevron must hug their content while the text column
        // absorbs the remaining width. Without this, UIStackView distributes
        // space equally and the badge stretches across the card — verified on
        // a real native host before this was added.
        badgeLabel.setContentCompressionResistancePriority(.required, for: .horizontal)
        badgeLabel.setContentHuggingPriority(.required, for: .horizontal)
        chevron.setContentCompressionResistancePriority(.required, for: .horizontal)
        chevron.setContentHuggingPriority(.required, for: .horizontal)
        textColumn.setContentHuggingPriority(.defaultLow, for: .horizontal)
        titleLabel.setContentCompressionResistancePriority(.defaultLow, for: .horizontal)
        subtitleLabel.setContentCompressionResistancePriority(.defaultLow, for: .horizontal)

        card.addSubview(row)

        NSLayoutConstraint.activate([
            row.leadingAnchor.constraint(equalTo: card.leadingAnchor, constant: 12),
            row.trailingAnchor.constraint(equalTo: card.trailingAnchor, constant: -12),
            row.topAnchor.constraint(equalTo: card.topAnchor, constant: 9),
            row.bottomAnchor.constraint(equalTo: card.bottomAnchor, constant: -9),
        ])

        window.addSubview(card)

        // Stack upward from the bottom, newest lowest, like the Flutter column.
        let offset = CGFloat(liveToasts.count) * 58 + 12
        NSLayoutConstraint.activate([
            card.leadingAnchor.constraint(equalTo: window.leadingAnchor, constant: 8),
            card.trailingAnchor.constraint(equalTo: window.trailingAnchor, constant: -8),
            card.bottomAnchor.constraint(
                equalTo: window.safeAreaLayoutGuide.bottomAnchor,
                constant: -offset
            ),
        ])

        liveToasts.append(card)
        while liveToasts.count > maxVisible, let oldest = liveToasts.first {
            dismiss(oldest)
        }

        card.alpha = 0
        UIView.animate(withDuration: 0.14) { card.alpha = 1 }

        DispatchQueue.main.asyncAfter(deadline: .now() + visibleDuration) {
            dismiss(card)
        }
    }

    private static func dismiss(_ card: UIView) {
        guard let index = liveToasts.firstIndex(of: card) else { return }
        liveToasts.remove(at: index)
        UIView.animate(
            withDuration: 0.14,
            animations: { card.alpha = 0 },
            completion: { _ in card.removeFromSuperview() }
        )
    }

    private static func dismissAll() {
        liveToasts.forEach { card in
            card.removeFromSuperview()
        }
        liveToasts.removeAll()
    }

    private static func keyWindow() -> UIWindow? {
        let scenes = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
        return scenes
            .first { $0.activationState == .foregroundActive }?
            .windows
            .first { $0.isKeyWindow }
            ?? scenes.first?.windows.first
    }

    private static func onMain(_ action: @escaping () -> Void) {
        if Thread.isMainThread {
            action()
        } else {
            DispatchQueue.main.async(execute: action)
        }
    }
}

/// The card view, which opens the dashboard when tapped.
private final class ToastCard: UIView {

    private let tab: String

    init(accent: UIColor, tab: String) {
        self.tab = tab
        super.init(frame: .zero)
        addGestureRecognizer(
            UITapGestureRecognizer(target: self, action: #selector(handleTap))
        )
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) is not used")
    }

    @objc private func handleTap() {
        removeFromSuperview()
        StackerDashboardPresenter.shared.present(initialTab: tab)
    }
}

/// A label with internal padding, for the status badge.
private final class PaddedLabel: UILabel {

    private let inset = UIEdgeInsets(top: 2, left: 6, bottom: 2, right: 6)

    override func drawText(in rect: CGRect) {
        super.drawText(in: rect.inset(by: inset))
    }

    override var intrinsicContentSize: CGSize {
        let base = super.intrinsicContentSize
        return CGSize(
            width: base.width + inset.left + inset.right,
            height: base.height + inset.top + inset.bottom
        )
    }
}

File: android/src/main/kotlin/com/stacker/stacker/StackerToast.kt
package com.stacker.stacker

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Application
import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import java.lang.ref.WeakReference

/**
 * Shows a per-call toast in a **native** host app, styled to match the Flutter
 * dashboard.
 *
 * In a Flutter app the toast stack is drawn by `StackerOverlay`. A native host
 * has no Flutter widget tree on screen — the engine only runs while the
 * dashboard activity is open — so that overlay can never appear.
 *
 * A platform [android.widget.Toast] was the obvious shortcut, but it looks
 * nothing like the Flutter version: a grey system pill with the system font,
 * no status colour, and not tappable. Since the same developer may see both on
 * the same project, the two must be visually the same component. This draws a
 * custom view into the foreground activity's content root instead, reusing the
 * exact colours, radii, and type scale from `StackerTheme` and `_ToastCard`.
 *
 * Enabled via [StackerAndroid.enable]; a no-op until then.
 */
object StackerToast {

    /** Toast policy, mirroring `ToastPolicy` on the Dart side. */
    enum class Policy {
        /** A toast for every completed call. */
        ALL,

        /** Only non-2xx responses and transport failures. */
        ERRORS_ONLY,

        /** No toasts. */
        NONE,
    }

    // Design tokens copied from StackerTheme so the two stay identical.
    // The card paints its own dark surface, so the *dark* variant of each
    // status colour is the correct one regardless of the host's theme.
    private const val SURFACE = 0xF01B2029.toInt()
    private const val SUCCESS = 0xFF4ADE80.toInt()
    private const val REDIRECT = 0xFFA78BFA.toInt()
    private const val WARN = 0xFFFBBF24.toInt()
    private const val ERROR = 0xFFF87171.toInt()
    private const val NEUTRAL = 0xFF94A3B8.toInt()

    /** How long a toast stays on screen, matching the Flutter default. */
    private const val VISIBLE_MS = 3_000L

    /** Bursts are coalesced so parallel requests cannot cover the screen. */
    private const val MAX_VISIBLE = 3

    private val mainHandler = Handler(Looper.getMainLooper())

    @SuppressLint("StaticFieldLeak")
    @Volatile
    private var appContext: Context? = null

    @Volatile
    private var policy: Policy = Policy.NONE

    /**
     * The activity currently in the foreground.
     *
     * Weak so a backgrounded activity is never retained — this class outlives
     * every activity in the process.
     */
    private var currentActivity: WeakReference<Activity>? = null

    private val liveToasts = mutableListOf<View>()

    /** Whether toasts are currently enabled. */
    @JvmStatic
    val isEnabled: Boolean
        get() = policy != Policy.NONE && appContext != null

    /** Configures the emitter. Called by [StackerAndroid.enable]. */
    fun configure(context: Context, policy: Policy) {
        val app = context.applicationContext
        this.appContext = app
        this.policy = policy

        // The toast has to be added to whichever activity is in front, so the
        // lifecycle is tracked rather than requiring the host to pass one in.
        if (app is Application) {
            app.unregisterActivityLifecycleCallbacks(lifecycleCallbacks)
            app.registerActivityLifecycleCallbacks(lifecycleCallbacks)
        }
    }

    /** Turns toasts off and releases every reference. */
    fun disable() {
        policy = Policy.NONE
        (appContext as? Application)?.unregisterActivityLifecycleCallbacks(lifecycleCallbacks)
        mainHandler.post { dismissAll() }
        appContext = null
        currentActivity = null
    }

    /**
     * Shows a toast for a completed call, if policy allows it.
     *
     * @param statusCode the HTTP status, or `null` for a transport failure.
     */
    fun showApiCall(method: String, path: String, statusCode: Int?, durationMs: Long?) {
        val isError = statusCode == null || statusCode >= 400
        when (policy) {
            Policy.NONE -> return
            Policy.ERRORS_ONLY -> if (!isError) return
            Policy.ALL -> Unit
        }

        val badge = statusCode?.toString() ?: "ERR"
        val accent = accentFor(statusCode)
        val subtitle = buildString {
            append(if (statusCode == null) "transport failure" else "completed")
            durationMs?.let { append("  •  ${it} ms") }
        }

        post { show(badge, "$method $path", subtitle, accent, "api") }
    }

    /** Shows a toast for a captured crash. Always shown when enabled. */
    fun showCrash(errorType: String, fatal: Boolean) {
        if (policy == Policy.NONE) return
        val badge = if (fatal) "FATAL" else "WARN"
        val accent = if (fatal) ERROR else WARN
        post { show(badge, errorType, "tap to open the dashboard", accent, "crashes") }
    }

    /** Mirrors `StackerTheme.statusColor` for the dark surface. */
    private fun accentFor(statusCode: Int?): Int = when {
        statusCode == null -> ERROR
        statusCode in 200..299 -> SUCCESS
        statusCode in 300..399 -> REDIRECT
        statusCode in 400..499 -> WARN
        statusCode >= 500 -> ERROR
        else -> NEUTRAL
    }

    /**
     * Builds and attaches the card. Main thread only.
     *
     * Laid out to match `_ToastCard`: a rounded dark surface with a 1px accent
     * border, a tinted status badge, a monospace title, and a muted subtitle.
     */
    private fun show(
        badge: String,
        title: String,
        subtitle: String,
        accent: Int,
        tab: String,
    ) {
        val activity = currentActivity?.get() ?: return
        if (activity.isFinishing || activity.isDestroyed) return
        val root = activity.findViewById<ViewGroup>(android.R.id.content) ?: return

        val density = activity.resources.displayMetrics.density
        fun dp(value: Int) = (value * density).toInt()

        val card = LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            setPadding(dp(12), dp(9), dp(12), dp(9))
            background = GradientDrawable().apply {
                shape = GradientDrawable.RECTANGLE
                cornerRadius = dp(10).toFloat()
                setColor(SURFACE)
                // 0.55 alpha on the accent, as in the Flutter card.
                setStroke(dp(1), Color.argb(140, Color.red(accent), Color.green(accent), Color.blue(accent)))
            }
            elevation = dp(6).toFloat()
            isClickable = true
            setOnClickListener {
                dismiss(this)
                StackerActivity.launch(activity, tab)
            }
        }

        card.addView(
            TextView(activity).apply {
                text = badge
                setTextColor(accent)
                typeface = Typeface.MONOSPACE
                setTypeface(typeface, Typeface.BOLD)
                setTextSize(TypedValue.COMPLEX_UNIT_SP, 11f)
                setPadding(dp(6), dp(2), dp(6), dp(2))
                background = GradientDrawable().apply {
                    shape = GradientDrawable.RECTANGLE
                    cornerRadius = dp(4).toFloat()
                    // 0.16 alpha tint behind the badge.
                    setColor(Color.argb(41, Color.red(accent), Color.green(accent), Color.blue(accent)))
                }
            },
            LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
            ),
        )

        val textColumn = LinearLayout(activity).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(9), 0, dp(4), 0)
        }
        textColumn.addView(
            TextView(activity).apply {
                text = title
                setTextColor(Color.WHITE)
                typeface = Typeface.MONOSPACE
                setTextSize(TypedValue.COMPLEX_UNIT_SP, 12f)
                maxLines = 1
                ellipsize = android.text.TextUtils.TruncateAt.END
            },
        )
        if (subtitle.isNotEmpty()) {
            textColumn.addView(
                TextView(activity).apply {
                    text = subtitle
                    setTextColor(Color.argb(179, 255, 255, 255))
                    typeface = Typeface.MONOSPACE
                    setTextSize(TypedValue.COMPLEX_UNIT_SP, 10f)
                    maxLines = 1
                    ellipsize = android.text.TextUtils.TruncateAt.END
                },
            )
        }
        card.addView(
            textColumn,
            LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f),
        )

        card.addView(
            TextView(activity).apply {
                text = "›"
                setTextColor(Color.argb(138, 255, 255, 255))
                setTextSize(TypedValue.COMPLEX_UNIT_SP, 16f)
            },
        )

        // Stack upward from the bottom, newest lowest, like the Flutter column.
        val params = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT,
        ).apply {
            gravity = Gravity.BOTTOM
            leftMargin = dp(8)
            rightMargin = dp(8)
            bottomMargin = dp(12) + liveToasts.size * dp(58)
        }

        root.addView(card, params)
        liveToasts.add(card)

        while (liveToasts.size > MAX_VISIBLE) {
            dismiss(liveToasts.first())
        }

        card.alpha = 0f
        card.animate().alpha(1f).setDuration(140).start()
        mainHandler.postDelayed({ dismiss(card) }, VISIBLE_MS)
    }

    private fun dismiss(card: View) {
        if (!liveToasts.remove(card)) return
        card.animate().alpha(0f).setDuration(140).withEndAction {
            (card.parent as? ViewGroup)?.removeView(card)
        }.start()
    }

    private fun dismissAll() {
        liveToasts.toList().forEach { dismiss(it) }
    }

    private fun post(action: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            action()
        } else {
            mainHandler.post(action)
        }
    }

    /**
     * Tracks the foreground activity.
     *
     * The dashboard activity is skipped: it draws its own Flutter UI and a
     * toast on top of it would be redundant, exactly as `showLauncherBubble`
     * and `toastPolicy` are disabled in the dashboard's own engine.
     */
    private val lifecycleCallbacks = object : Application.ActivityLifecycleCallbacks {
        override fun onActivityResumed(activity: Activity) {
            if (activity is StackerActivity) return
            currentActivity = WeakReference(activity)
        }

        override fun onActivityPaused(activity: Activity) {
            if (currentActivity?.get() === activity) {
                dismissAll()
                currentActivity = null
            }
        }

        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) = Unit
        override fun onActivityStarted(activity: Activity) = Unit
        override fun onActivityStopped(activity: Activity) = Unit
        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) = Unit
        override fun onActivityDestroyed(activity: Activity) = Unit
    }
}
